### DISCUSSION TITLE
Recover public address from signature

### DISCUSSION THREAD

**Discussion by rodionlim** [ConsenSys/quorum#1540] [Q&A]
Hi, I understand that when signing a raw private transaction, we have to sign the transaction first before setting V to 37/38. Is it possible to derive the public key based off the signature (R, S, V) and rlp encoded raw transaction, i.e. via some sort of `ecrecover` method? If so, are there any examples on doing that?

Thank you.

**Comment by baptiste-b-pegasys** [comment]
Hello, are you using ethereumjs-tx?

There is a field for the sender pub key and a function for the access: https://github.com/ethereumjs/ethereumjs-tx/blob/5c81b38/src/transaction.ts#L221

I can see that there is ecrecover just after: https://github.com/ethereumjs/ethereumjs-tx/blob/5c81b38/src/transaction.ts#L244
That might be the example you are looking for.

What do you mean by public address ? because I can see it's about encryption key, I am confused.

**Comment by rodionlim** [comment]
Hi, sorry, I meant deriving the sender pub key from the r, s and v values and original transaction. Yes, I am trying to use something similar that you have highlighted in your comments, but I'm unable to derive the right public key. I'm not sure if the recovery is failing because of the setting of V to 37/38, so I was wondering if this was something that anyone has tried before with a working example.

**Reply by baptiste-b-pegasys** [reply]
Private transactions have a specific addition of 37 on V

```
func (qs QuorumPrivateTxSigner) SignatureValues(tx *Transaction, sig []byte) (R, S, V *big.Int, err error) {
	r, s, _, err := HomesteadSigner{}.SignatureValues(tx, sig)
	// update v for private transaction marker: needs to be 37 (0+37) or 38 (1+37) for a private transaction.
	v := new(big.Int).SetBytes([]byte{sig[64] + 37})
	return r, s, v, nil
}
```

I think it would be a good idea to set back to 0 or 1 before using ecrecover.
I have some issue with my node/npm. I'll let you know if I am able to test that.

I want to do that:
```
const EthereumTx = require('ethereumjs-tx');
const txData = {
  nonce: '0x00',
  gasPrice: '0x0',
  gasLimit: '0x2710',
  to: '0x0000000000000000000000000000000000000000',
  value: '0x00',
  data: '0x9dfaff81059520dac4844f9c6cb71da45a246fa1e05a11d22dbb8693e813dc235da812dfa313fba6095e3a10ebefca8607e5eb18085c917de6a51b693a7c6d99',
  v: '0x26',
  r: '0x64c7cd5d840a14f9eeeb52870dafbb0e277c4eb647d47707460ef2d3f8137d28',
  s: '0x6a7ff69fd3a3788b54a6fd5a21ea5ca31d1517c68b9785cc058709720aae319f'
};
const tx = new Transaction(txData);
echo(tx.getSenderAddress());
```

**Reply by rodionlim** [reply]
Exactly, that's what I was trying to figure out, i.e. how to derive the sender address from the signed transaction with a V that was replaced with 37/38

**Reply by baptiste-b-pegasys** [reply]
There is no need to change the V value from what I am seeing.

V = 0 or 1 is only for mainnet

```
node main.js
Address {
  buf: <Buffer b9 3a c0 2a b3 be 05 e9 25 8a 28 e7 3d ce 2f 37 92 ed cd 4d>
}
```

[Archive.zip](https://github.com/ConsenSys/quorum/files/9830852/Archive.zip)

**Reply by rodionlim** [reply]
Hey @baptiste-b-pegasys, can you provide me the snippet of code that you used to get the v, r and s? And the corresponding private key that you used to sign the transaction. I could recover an address on my end, the only problem I had was it didn't tie with the private key that I signed the transaction with.

**Reply by baptiste-b-pegasys** [reply]
OK I see now, I will adapt this library then in order to test the 0 value.
The function returns the public key, if it returns the private key, I would be worried I think.

**Reply by rodionlim** [reply]
I think you misunderstood me, it returns a public key for me, just not the one that signed the transaction based off the private key that I used. I'm guessing that I am using the wrong parameters to do a ecrecover in my code.
